### PR TITLE
RevEng: Hook up IApplicationShutdown Cancel event and use async methods where possible

### PR DESCRIPTION
**Pull request by lajones**
@AndriySvyryd @divega Updating as agreed. Using .GetAwaiter().GetResult() instead of .Result returns the underlying exception rather than an AggregateException if an exception is thrown.

### CODE REVIEW

File: src/EntityFramework.Commands/DatabaseTool.cs [hunk 1]
```diff
@@ -70,7 +70,7 @@ public class DatabaseTool
             };
 
             var generator = new ReverseEngineeringGenerator(_serviceProvider);
-            return generator.GenerateAsync(configuration).Result;
+            return generator.GenerateAsync(configuration).GetAwaiter().GetResult();
```
Comment (divega) [inline src/EntityFramework.Commands/DatabaseTool.cs]: Why not make things async throughout? AFAIK DNX supports async `Program.Main`. Is it hard to use async methods from PowerShell?

cc @bricelam

**Comment by lajones** [comment]
@danielmarbach Note: this addresses the changes you included in #2240 as well.

**Comment by danielmarbach** [comment]
Ok. Yeah didn't have time to do it. Btw. There is a [NotNull] missing in the Program

**Comment by lajones** [comment]
Thanks @danielmarbach - well-spotted. Will fix.

File: src/EntityFramework.Commands/Program.cs [hunk 2]
```diff
@@ -222,7 +232,11 @@ public virtual int Main([NotNull] string[] args)
                             "[connectionString]",
                             "The connection string of the database");
 
-                    revEng.OnExecute(() => ReverseEngineer(connectionString.Value));
+                    revEng.OnExecute(async () =>
+                        {
+                            return await ReverseEngineerAsync(
+                                connectionString.Value, _applicationShutdown.ShutdownRequested);
+                        });
```
Comment (lajones) [inline src/EntityFramework.Commands/Program.cs]: Note: Have updated the above to simpler async () => await ... syntax.

File: src/EntityFramework.Commands/DatabaseTool.cs [hunk 3]
```diff
@@ -70,7 +73,7 @@ public class DatabaseTool
             };
 
             var generator = new ReverseEngineeringGenerator(_serviceProvider);
-            return generator.GenerateAsync(configuration).Result;
+            return await generator.GenerateAsync(configuration, cancellationToken);
```
Comment (AndriySvyryd) [inline src/EntityFramework.Commands/DatabaseTool.cs]: No need to use await here, just return what GenerateAsync returns

File: src/EntityFramework.Commands/Executor.cs [hunk 4]
```diff
@@ -249,13 +251,18 @@ public ReverseEngineer([NotNull] Executor executor, [NotNull] object resultHandl
...
 
-        public virtual IEnumerable<string> ReverseEngineerImpl(
-            [NotNull] string providerAssemblyName, [NotNull] string connectionString) =>
-                _databaseTool.ReverseEngineer(providerAssemblyName, connectionString, _rootNamespace, _projectDir);
+        public virtual async Task<IEnumerable<string>> ReverseEngineerImplAsync(
+            [NotNull] string providerAssemblyName,
+            [NotNull] string connectionString,
+            CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await _databaseTool.ReverseEngineerAsync(
```
Comment (AndriySvyryd) [inline src/EntityFramework.Commands/Executor.cs]: remove await

File: src/EntityFramework.Commands/Program.cs (hunk 2 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Commands/Program.cs]: You can remove async/await

**Comment by AndriySvyryd** [comment]
:shipit:

File: src/EntityFramework.Commands/Executor.cs [hunk 5]
```diff
@@ -249,13 +251,18 @@ public ReverseEngineer([NotNull] Executor executor, [NotNull] object resultHandl
 
                 var providerAssemblyName = DatabaseTool._defaultReverseEngineeringProviderAssembly;
 
-                Execute(() => executor.ReverseEngineerImpl(providerAssemblyName, connectionString));
+                Execute(() => executor.ReverseEngineerImplAsync(providerAssemblyName, connectionString).GetAwaiter().GetResult());
```
Comment (divega) [inline src/EntityFramework.Commands/Executor.cs]: Can this be made async?

Comment (divega) [inline src/EntityFramework.Commands/Executor.cs]: Or is this the PowerShell stuff?

Comment (divega) [inline src/EntityFramework.Commands/Executor.cs]: I guess it is :smile:

File: src/EntityFramework.Commands/DatabaseTool.cs (hunk 3 above)
Comment (lajones) [inline src/EntityFramework.Commands/DatabaseTool.cs]: If you do that the compiler complains that this method has no await in it.

File: src/EntityFramework.Commands/Executor.cs (hunk 4 above)
Comment (lajones) [inline src/EntityFramework.Commands/Executor.cs]: As above

File: src/EntityFramework.Commands/Executor.cs (hunk 5 above)
Comment (lajones) [inline src/EntityFramework.Commands/Executor.cs]: Yep - this is the place where we agreed the sync-async crossover would take place for Powershell.

File: src/EntityFramework.Commands/DatabaseTool.cs (hunk 3 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Commands/DatabaseTool.cs]: you need to remove the async keyword as well

File: src/EntityFramework.Commands/Program.cs (hunk 2 above)
Comment (lajones) [inline src/EntityFramework.Commands/Program.cs]: Nope - produces compiler errors warning that the async and await are necessary.

Comment (divega) [inline src/EntityFramework.Commands/Program.cs]: I suspect in this case it is needed.

**Comment by lajones** [comment]
OK. Have talked with Andriy and sorted out the async calls.

**Comment by lajones** [comment]
Fixed with commit c75abcf.

**Pull request closed without merging**